packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -1471,11 +1418,199 @@ class RenderPhysicalModel extends _RenderCustomClip<RRect> {
     }
   }
 
+  @override
+  bool hitTest(HitTestResult result, { Offset position }) {
+    if (_clipper != null) {
+      _updateClip();
+      assert(_clip != null);
+      if (!_clip.contains(position))
+        return false;
+    }
+    return super.hitTest(result, position: position);
+  }
+
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder description) {
     super.debugFillProperties(description);
     description.add(new DiagnosticsProperty<BoxShape>('shape', shape));
     description.add(new DiagnosticsProperty<BorderRadius>('borderRadius', borderRadius));
+  }
+}
+
+/// Creates a physical model layer that clips its children to a [ShapeBorder].
+///
+/// A physical model layer casts a shadow based on its [elevation].
+///
+/// See also:
+///
+/// * [RenderPhysicalModel], which is optimized for rounded rectangles and
+///   circles.
+class RenderPhysicalShape extends _RenderPhysicalModelBase<Path> {
+  /// Creates an arbitrary shape clip.
+  ///
+  /// The [color], [shape], and [textDirection] are required.
+  ///
+  /// The [shape], [elevation], [color], [textDirection] and [shadowColor] must
+  /// not be null.
+  RenderPhysicalShape({
+    RenderBox child,
+    @required ShapeBorder shape,
+    double elevation: 0.0,
+    @required Color color,
+    Color shadowColor: const Color(0xFF000000),
+    @required TextDirection textDirection,
+  }) : assert(shape != null),
+       assert(elevation != null),
+       assert(color != null),
+       assert(shadowColor != null),
+       assert(textDirection != null),
+       _shape = shape,
+       _textDirection = textDirection,
+       super(
+         child: child,
+         elevation: elevation,
+         color: color,
+         shadowColor: shadowColor
+       );
+
+  /// The shape of the layer.
+  ShapeBorder get shape => _shape;
+  ShapeBorder _shape;
+  set shape(ShapeBorder value) {
+    assert(value != null);
+    if (shape == value)
+      return;
+    _shape = value;
+    _markNeedsClip();
+  }
+
+  /// The text direction to use for getting the outer path for [shape].
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    assert(value != null);
+    if (textDirection == value)
+      return;
+    _textDirection = value;
+    _markNeedsClip();
+  }
+
+  @override
+  Path get _defaultClip {
+    assert(hasSize);
+    return shape.getOuterPath(Offset.zero & size, textDirection: _textDirection);
```

Comment:
As mentioned above, I agree the inner/outer paths are a bit ambiguous here as we don't draw a border, but just clip the shape.
I guess if we will be drawing a border, then it might or might not make sense to clip to the outer path (if this render object or a child render object are the ones drawing the border we probably want to clip to the outer path, if a parent render object is the one drawing the border than it makes sense to clip to the inner path).